### PR TITLE
Remove assignment and declarations in if statements

### PR DESCRIPTION
**Pull request by hpvb**
After discussing with @reduz and @akien-mga it was decided that we do
not allow assignments or declarations in if statements. This PR removes
the instances of this I could find by automated means.

### CODE REVIEW

**Comment by raymoo** [comment]
Even for guaranteed initialization purposes?

**Comment by raymoo** [comment]
I think that because declarations in if conditions are syntactically and semantically distinct from simply using an assignment expression in the condition, they should be considered separately.

**Comment by NathanWarden** [comment]
100% agree with this pr. Thank you! :)

File: editor/import/resource_importer_scene.cpp [hunk 1]
```diff
@@ -230,8 +230,8 @@ Node *ResourceImporterScene::_fix_node(Node *p_node, Node *p_root, Map<Ref<Array
 
 		if (isroot)
 			return p_node;
-
-		if (MeshInstance *mi = Object::cast_to<MeshInstance>(p_node)) {
```
Comment (raymoo) [inline editor/import/resource_importer_scene.cpp]: This is a declaration / definition, not an assignment.

File: editor/plugins/canvas_item_editor_plugin.cpp [hunk 2]
```diff
@@ -580,7 +580,8 @@ void CanvasItemEditor::_key_move(const Vector2 &p_dir, bool p_snap, KeyMoveMODE
 
 		} else { // p_move_mode==MOVE_LOCAL_BASE || p_move_mode==MOVE_LOCAL_WITH_ROT
 
-			if (Node2D *node_2d = Object::cast_to<Node2D>(canvas_item)) {
```
Comment (raymoo) [inline editor/plugins/canvas_item_editor_plugin.cpp]: Ditto

File: editor/plugins/canvas_item_editor_plugin.cpp [hunk 3]
```diff
@@ -589,9 +590,10 @@ void CanvasItemEditor::_key_move(const Vector2 &p_dir, bool p_snap, KeyMoveMODE
 				}
 				node_2d->set_position(node_2d->get_position() + drag);
 
-			} else if (Control *control = Object::cast_to<Control>(canvas_item)) {
```
Comment (raymoo) [inline editor/plugins/canvas_item_editor_plugin.cpp]: Ditto

File: editor/scene_tree_dock.cpp [hunk 4]
```diff
@@ -582,13 +582,13 @@ void SceneTreeDock::_tool_selected(int p_tool, bool p_confirm_override) {
 
 			new_scene_from_dialog->popup_centered_ratio();
 			new_scene_from_dialog->set_title(TTR("Save New Scene As.."));
-
 		} break;
 		case TOOL_COPY_NODE_PATH: {
 			List<Node *> selection = editor_selection->get_selected_node_list();
-
-			if (List<Node *>::Element *e = selection.front()) {
```
Comment (raymoo) [inline editor/scene_tree_dock.cpp]: Ditto

File: editor/scene_tree_dock.cpp [hunk 5]
```diff
@@ -597,8 +597,10 @@ void SceneTreeDock::_tool_selected(int p_tool, bool p_confirm_override) {
 		} break;
 		case TOOL_SCENE_EDITABLE_CHILDREN: {
 			List<Node *> selection = editor_selection->get_selected_node_list();
-			if (List<Node *>::Element *e = selection.front()) {
```
Comment (raymoo) [inline editor/scene_tree_dock.cpp]: Ditto

File: editor/scene_tree_dock.cpp [hunk 6]
```diff
@@ -617,8 +619,10 @@ void SceneTreeDock::_tool_selected(int p_tool, bool p_confirm_override) {
 		} break;
 		case TOOL_SCENE_USE_PLACEHOLDER: {
 			List<Node *> selection = editor_selection->get_selected_node_list();
-			if (List<Node *>::Element *e = selection.front()) {
```
Comment (raymoo) [inline editor/scene_tree_dock.cpp]: Ditto

File: editor/scene_tree_dock.cpp [hunk 7]
```diff
@@ -635,15 +639,16 @@ void SceneTreeDock::_tool_selected(int p_tool, bool p_confirm_override) {
 		} break;
 		case TOOL_SCENE_CLEAR_INSTANCING: {
 			List<Node *> selection = editor_selection->get_selected_node_list();
-			if (List<Node *>::Element *e = selection.front()) {
```
Comment (raymoo) [inline editor/scene_tree_dock.cpp]: Ditto

File: editor/scene_tree_dock.cpp [hunk 8]
```diff
@@ -656,8 +661,10 @@ void SceneTreeDock::_tool_selected(int p_tool, bool p_confirm_override) {
 		} break;
 		case TOOL_SCENE_OPEN: {
 			List<Node *> selection = editor_selection->get_selected_node_list();
-			if (List<Node *>::Element *e = selection.front()) {
```
Comment (raymoo) [inline editor/scene_tree_dock.cpp]: Ditto

File: editor/scene_tree_dock.cpp [hunk 9]
```diff
@@ -667,8 +674,10 @@ void SceneTreeDock::_tool_selected(int p_tool, bool p_confirm_override) {
 		} break;
 		case TOOL_SCENE_CLEAR_INHERITANCE_CONFIRM: {
 			List<Node *> selection = editor_selection->get_selected_node_list();
-			if (List<Node *>::Element *e = selection.front()) {
```
Comment (raymoo) [inline editor/scene_tree_dock.cpp]: Ditto

File: editor/scene_tree_dock.cpp [hunk 10]
```diff
@@ -677,8 +686,10 @@ void SceneTreeDock::_tool_selected(int p_tool, bool p_confirm_override) {
 		} break;
 		case TOOL_SCENE_OPEN_INHERITED: {
 			List<Node *> selection = editor_selection->get_selected_node_list();
-			if (List<Node *>::Element *e = selection.front()) {
```
Comment (raymoo) [inline editor/scene_tree_dock.cpp]: Ditto

File: editor/scene_tree_dock.cpp [hunk 11]
```diff
@@ -677,8 +686,10 @@ void SceneTreeDock::_tool_selected(int p_tool, bool p_confirm_override) {
...
-				if (Node *node = e->get()) {
```
Comment (raymoo) [inline editor/scene_tree_dock.cpp]: Ditto

File: scene/3d/bone_attachment.cpp [hunk 12]
```diff
@@ -71,7 +71,8 @@ void BoneAttachment::_get_property_list(List<PropertyInfo> *p_list) const {
 
 void BoneAttachment::_check_bind() {
 
-	if (Skeleton *sk = Object::cast_to<Skeleton>(get_parent())) {
```
Comment (raymoo) [inline scene/3d/bone_attachment.cpp]: This is a declaration / definition, not an assignment.

File: scene/3d/bone_attachment.cpp [hunk 13]
```diff
@@ -86,7 +87,8 @@ void BoneAttachment::_check_unbind() {
 
 	if (bound) {
 
-		if (Skeleton *sk = Object::cast_to<Skeleton>(get_parent())) {
```
Comment (raymoo) [inline scene/3d/bone_attachment.cpp]: This is a declaration / definition, not an assignment.

File: scene/3d/collision_shape.cpp [hunk 14]
```diff
@@ -50,7 +50,8 @@ void CollisionShape::make_convex_from_brothers() {
 	for (int i = 0; i < p->get_child_count(); i++) {
 
 		Node *n = p->get_child(i);
-		if (MeshInstance *mi = Object::cast_to<MeshInstance>(n)) {
```
Comment (raymoo) [inline scene/3d/collision_shape.cpp]: This is a declaration / definition, not an assignment.

File: scene/3d/gi_probe.cpp [hunk 15]
```diff
@@ -1113,7 +1113,8 @@ void GIProbe::_find_meshes(Node *p_at_node, Baker *p_baker) {
 		}
 	}
 
-	if (Spatial *s = Object::cast_to<Spatial>(p_at_node)) {
```
Comment (raymoo) [inline scene/3d/gi_probe.cpp]: This is a declaration / definition, not an assignment.

File: scene/3d/interpolated_camera.cpp [hunk 16]
```diff
@@ -55,8 +55,8 @@ void InterpolatedCamera::_notification(int p_what) {
 				Transform local_transform = get_global_transform();
 				local_transform = local_transform.interpolate_with(target_xform, delta);
 				set_global_transform(local_transform);
-
-				if (Camera *cam = Object::cast_to<Camera>(node)) {
```
Comment (raymoo) [inline scene/3d/interpolated_camera.cpp]: This is a declaration / definition, not an assignment.

File: scene/gui/file_dialog.cpp [hunk 17]
```diff
@@ -183,8 +183,8 @@ void FileDialog::_action_pressed() {
 		String path = dir_access->get_current_dir();
 
 		path = path.replace("\\", "/");
-
-		if (TreeItem *item = tree->get_selected()) {
```
Comment (raymoo) [inline scene/gui/file_dialog.cpp]: This is a declaration / definition, not an assignment.

**Review by raymoo — CHANGES_REQUESTED** [review]
Declaration + Definition =/= assignment

The "declaration if" is a separate construct from just using assignment operators in the condition, and helps prevent null pointer dereferencing. It does not cause reading problems like using assignments in the condition does, because it is clearly visible and you cannot mix it with other conditions.

**Comment by hpvb** [comment]
@raymoo I made the exact same argument but @reduz and @akien-mga didn't want either case. I'll update the PR's title.

**Comment by raymoo** [comment]
Ok, I guess it's their call in the end.

**Comment by raymoo** [comment]
Another option that avoids a null dereference later on is to put an extra pair of braces around the whole thing. For example:

```
if (Banana* banana = GetABanana()) {
  // banana is not null in here
}

// becomes
{
  Banana* banana = GetABanana()
  if (banana) {
    // banana is not null
  }
}
```

**Pull request merged**